crates/bevy_ui/src/keyboard_navigation.rs

```diff
@@ -0,0 +1,173 @@
+use crate::{Interaction, UiStack};
+use bevy_a11y::Focus;
+use bevy_ecs::prelude::Component;
+use bevy_ecs::query::With;
+use bevy_ecs::system::{Local, Query, Res, ResMut};
+use bevy_ecs::{change_detection::DetectChangesMut, entity::Entity};
+use bevy_input::{prelude::KeyCode, Input};
+use bevy_reflect::{FromReflect, Reflect};
+use bevy_render::view::ComputedVisibility;
+
+/// A component that represents if a UI element is focused.
+#[derive(Reflect, FromReflect, Component, Clone, Debug, Default, Eq, PartialEq)]
+pub struct Focusable {
+    focus_state: FocusState,
+}
+
+impl Focusable {
+    /// The entity is currently focused, similar to the `:focus` css pseudo-class.
+    /// To check if the focus has been achieved through keyboard navigation, see [`Focusable::is_focus_visible`].
+    pub fn is_focused(&self) -> bool {
+        matches!(self.focus_state, FocusState::Focused { .. })
+    }
+
+    /// Focus has been reached through keyboard navigation and so a focus style should be displayed.
+    /// This is similar to the `:focus-visible` pseudo-class in css.
+    pub fn is_focus_visible(&self) -> bool {
+        matches!(self.focus_state, FocusState::Focused { visible: true })
+    }
+}
+
+#[derive(Reflect, FromReflect, Component, Clone, Debug, Default, Eq, PartialEq)]
+enum FocusState {
+    /// Entity is not focused
+    #[default]
+    None,
+    /// Entity is focused
+    Focused {
+        /// Focus has been reached through keyboard navigation and so a focus style should be displayed.
+        /// This is similar to the `:focus-visible` pseudo-class in css.
+        visible: bool,
+    },
+}
+
+/// Should the [`keyboard_navigation_system`] run?
+pub(crate) fn tab_pressed(keyboard_input: Res<Input<KeyCode>>) -> bool {
+    keyboard_input.just_pressed(KeyCode::Tab)
+}
```

Comment:
I know full input mapping is out of scope for this PR, but there should be at least a way to opt-out of the behavior, so that bevy users can use `tab`.

Think of a game like minecraft where you open the inventory by pressing the `tab` key. You wouldn't want that to affect the position of the cursor.

There should be a `Resource` (for example: `KeyboardInput { pub enabled: bool }`) and users should be able to toggle it. Then, navigation systems execution should be conditioned on this resource.

I think it's a major requirement. You really want to be able to condition navigation on specific game states.

Reply:
Agreed, I'll add that.